client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java

```diff
@@ -518,6 +523,33 @@ public <K, V> ShuffleWriter<K, V> getWriter(
         shuffleHandleInfo);
   }
 
+  /**
+   * Provides a task attempt id that is unique for a shuffle stage.
+   *
+   * <p>We are not using context.taskAttemptId() here as this is a monotonically increasing number
+   * that is unique across the entire Spark app which can reach very large numbers, which can
+   * practically reach LONG.MAX_VALUE. That would overflow the bits in the block id.
+   *
+   * <p>Here we use the map index or task id, appended by the attempt number per task. The map index
+   * is limited by the number of partitions of a stage. The attempt number per task is limited /
```

Comment:
Attempt number may be larger than maxFailures.
If we fail 3 times, when fourth attempt run, Spark may trigger a speculative task at the same time. We will have 5 attempts.

Reply:
That is a bit surprising, but looking at the relevant code, the max failure is not considered when resubmitting a task as speculative:
https://github.com/apache/spark/blob/2abd3a2f445e86337ad94da19f301cb2b8bc232f/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L1226-L1227

I will account for that!

Reply:
Good catch, accounted for: dcd03bc